app/models/cost_query/cache.rb

```diff
@@ -0,0 +1,75 @@
+#-- copyright
+# OpenProject Reporting Plugin
+#
+# Copyright (C) 2010 - 2014 the OpenProject Foundation (OPF)
+#
+# This program is free software; you can redistribute it and/or
+# modify it under the terms of the GNU General Public License
+# version 3.
+#
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with this program; if not, write to the Free Software
+# Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+#++
+
+module CostQuery::Cache
+  class << self
+
+    def check
+      reset! if reset_required?
+    end
+
+    def reset!
+      CostQuery::Filter.reset!
+      CostQuery::Filter::CustomFieldEntries.reset!
+      CostQuery::GroupBy.reset!
+      CostQuery::GroupBy::CustomFieldEntries.reset!
+
+      update_reset_on
+    end
+
+    protected
+
+    attr_accessor :custom_fields_updated_on,
+                  :custom_fields_id_sum
+
+    def invalid?
+      updated_on = fetch_custom_field_updated_at
+      id_sum = fetch_custom_fields_changed
+
+      custom_fields_updated_on != updated_on ||
+        custom_fields_id_sum != id_sum
+    end
+
+    def update_reset_on
+      return if caching_disabled?
+
+      self.custom_fields_updated_on = fetch_custom_field_updated_at
+      self.custom_fields_id_sum = fetch_custom_fields_changed
+    end
+
+    def fetch_custom_field_updated_at
+      WorkPackageCustomField.maximum(:updated_at)
+    end
+
+    def fetch_custom_fields_changed
+      WorkPackageCustomField.sum(:id) + WorkPackageCustomField.count
```

Comment:
Is this supposed to be a quick hash algorithm that can be used to determine whether custom fields were added/removed?

Caches are usually designed to "better be safe than sorry", that means it is okay to rebuild the cache more often than needed as long as the data is **always** valid. Because you don't want to debug errors that are caused by invalid caches.

I have yet to prove that your algorithm can fail to detect a change, admittedly I thought I would come up with a counter example immediately, but I did fail to do that. Stay tuned...